hivemind/p2p/p2p_daemon_bindings/control.py

```diff
@@ -0,0 +1,217 @@
+"""
+Originally taken from: https://github.com/mhchia/py-libp2p-daemon-bindings
+Licence: MIT
+Author: Kevin Mai-Husan Chia
+"""
+
+import logging
+from typing import AsyncIterator, Awaitable, Callable, Dict, Iterable, Sequence, Tuple
+
+import asyncio
+from contextlib import asynccontextmanager
+from multiaddr import Multiaddr, protocols
+from hivemind.p2p.p2p_daemon_bindings.datastructures import PeerInfo, StreamInfo, ID
+from hivemind.proto import p2pd_pb2 as p2pd_pb
+from hivemind.p2p.p2p_daemon_bindings.utils import DispatchFailure, read_pbmsg_safe, write_pbmsg, raise_if_failed
+
+StreamHandler = Callable[[StreamInfo, asyncio.StreamReader, asyncio.StreamWriter], Awaitable[None]]
+
+_supported_conn_protocols = (
+    protocols.P_IP4,
+    # protocols.P_IP6,
+    protocols.P_UNIX,
+)
+
+
+def parse_conn_protocol(maddr: Multiaddr) -> int:
+    proto_codes = set(proto.code for proto in maddr.protocols())
+    proto_cand = proto_codes.intersection(_supported_conn_protocols)
+    if len(proto_cand) != 1:
+        supported_protos = (
+            protocols.protocol_with_code(proto) for proto in _supported_conn_protocols
+        )
+        raise ValueError(
+            f"connection protocol should be only one protocol out of {supported_protos}"
+            f", maddr={maddr}"
+        )
+    return tuple(proto_cand)[0]
+
+
+class DaemonConnector:
+    control_maddr: Multiaddr
+    logger = logging.getLogger("p2pclient.DaemonConnector")
+    DEFAULT_CONTROL_MADDR = "/unix/tmp/p2pd.sock"
+
+    def __init__(self, control_maddr: Multiaddr = None) -> None:
+        if control_maddr is None:
+            control_maddr = Multiaddr(self.DEFAULT_CONTROL_MADDR)
+        self.control_maddr = control_maddr
+
+    async def open_connection(self) -> (asyncio.StreamReader, asyncio.StreamWriter):
+        proto_code = parse_conn_protocol(self.control_maddr)
+        if proto_code == protocols.P_UNIX:
+            control_path = self.control_maddr.value_for_protocol(protocols.P_UNIX)
+            self.logger.debug(
+                "DaemonConnector %s opens connection to %s", self, self.control_maddr
+            )
+            return await asyncio.open_unix_connection(control_path)
+        elif proto_code == protocols.P_IP4:
+            host = self.control_maddr.value_for_protocol(protocols.P_IP4)
+            port = int(self.control_maddr.value_for_protocol(protocols.P_TCP))
+            return await asyncio.open_connection(host, port)
+        else:
+            raise ValueError(
+                f"protocol not supported: protocol={protocols.protocol_with_code(proto_code)}"
+            )
+
+
+class ControlClient:
+    listen_maddr: Multiaddr
+    daemon_connector: DaemonConnector
+    handlers: Dict[str, StreamHandler]
+    logger = logging.getLogger("p2pclient.ControlClient")
+    DEFAULT_LISTEN_MADDR = "/unix/tmp/p2pclient.sock"
+
+    def __init__(
+        self, daemon_connector: DaemonConnector, listen_maddr: Multiaddr = None
+    ) -> None:
```

Comment:
Are we hitting the 120 column limit with this type annotation? Better if it's a single line

Reply:
With default listen_maddr = Multiaddr(DEFAULT_LISTEN_MADDR) it is longer than 120